### PR TITLE
ENH: Add annotations for `np.core.arrayprint`

### PR DESCRIPTION
**Pull request by BvB93**
Per the title: this PR adds annotations for all array-printing-related functions in `np.core.arrayprint`.

@charris continuing from https://github.com/numpy/numpy/pull/7859#issuecomment-747588637: 
The signature of `np.set_printoptions` can quite easily be annotated as it turns out,
so annotations can definitely help here with aforementioned issue.

### CODE REVIEW

File: numpy/core/arrayprint.pyi [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    timedelta: Callable[[timedelta64], str]
+    datetime: Callable[[datetime64], str]
+    float: Callable[[floating[Any]], str]
+    longfloat: Callable[[longdouble], str]
+    complexfloat: Callable[[complexfloating[Any, Any]], str]
+    longcomplexfloat: Callable[[clongdouble], str]
+    void: Callable[[void], str]
+    numpystr: Callable[[_CharLike], str]
+    object: Callable[[object], str]
+    str: Callable[[Any], str]  # Unused but still present?
```
Comment (BvB93) [inline numpy/core/arrayprint.pyi:43]: The docstring of `np.setprintoptions` ([doc](https://github.com/numpy/numpy/blob/bac54ecb0b815eea2e8116d21aa96e1387793468/numpy/core/arrayprint.py#L132-L157)) mentions an optional `str` key, though for the past 8 years it has actually been non-functional (see https://github.com/numpy/numpy/pull/459).

Can we just remove any mention of it, both here and in the docstring?

Comment (rgommers) [reply numpy/core/arrayprint.pyi:43]: Sounds fine to me.

File: numpy/core/arrayprint.pyi [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    threshold: Optional[int] = ...,
+    edgeitems: Optional[int] = ...,
+    linewidth: Optional[int] = ...,
+    suppress: Optional[bool] = ...,
+    nanstr: Optional[str] = ...,
+    infstr: Optional[str] = ...,
+    formatter: Optional[_FormatDict] = ...,
+    sign: Optional[Literal["-", "+", " "]] = ...,
+    floatmode: Optional[
+        Literal["fixed", "unique", "maxprec", "maxprec_equal"]
```
Comment (eric-wieser) [inline numpy/core/arrayprint.pyi:74]: This might be worth abstracting to a `_FloatMode =  Literal["fixed", "unique", "maxprec", "maxprec_equal"]`

Comment (BvB93) [reply numpy/core/arrayprint.pyi]: Not a bad idea, done as of https://github.com/numpy/numpy/pull/18032/commits/05fa6e4bc219fb4db0808ef5cf9aed96c8ec0844.

File: numpy/core/arrayprint.pyi (hunk 1 above)
Comment (BvB93) [reply numpy/core/arrayprint.pyi]: Done as of https://github.com/numpy/numpy/pull/18032/commits/82e5747dd40cbbde9c728452f21fa4716492c81d.

Just to stress: the above-mentioned push does not deprecate anything, 
it merely removes a few lines of unused code (and updates the docstring).

**Comment by charris** [comment]
Thanks Bas.

**Pull request merged**